src/Inciter/AMR/refinement.h

```diff
@@ -0,0 +1,779 @@
+#ifndef AMR_refinement_h
+#define AMR_refinement_h
+
+#include <algorithm>
+
+#include "tet_store.h"
+#include "node_connectivity.h"
+
+// TODO: make this have a base class to support multiple generator schemes
+// using the policy design pattern
+
+namespace AMR {
+
+    class refinement_t {
+        private:
+
+            const size_t DEFAULT_REFINEMENT_LEVEL = 0; //TODO: Is this in the right place?
+            const size_t MIN_REFINEMENT_LEVEL = DEFAULT_REFINEMENT_LEVEL;
+
+            // TODO: Tidy up edge store references here
+            tet_store_t* tet_store;
+            node_connectivity_t* node_connectivity;
+
+        public:
+
+            const size_t MAX_REFINEMENT_LEVEL = 4;
+
+            refinement_t(tet_store_t* ts, node_connectivity_t* ns) : tet_store(ts), node_connectivity(ns)
+            {
+            }
+
+            // TODO: Document this
+            child_id_list_t generate_child_ids(size_t parent_id, size_t count = MAX_CHILDREN)
+            {
+                //return morton_id_generator_t::get_children_ids(parent_id);
+                return tet_store->generate_child_ids(parent_id, count);
+            }
+
+            /**
+             * @brief function to detect when an invalid refinement is
+             * invoked
+             *
+             * @param tet_id Id the of the tet which will be refined
+             *
+             * @return A bool stating if the tet can be validly refined
+             */
+            bool check_allowed_refinement(size_t tet_id)
+            {
+                Refinement_State& master_element = tet_store->data(tet_id);
+
+                // These asserts mean we never actually try refine a 1:2 or 1:4
+                assert( master_element.refinement_case !=
+                        Refinement_Case::one_to_two);
+                assert( master_element.refinement_case !=
+                        Refinement_Case::one_to_four);
+
+                assert( tet_store->is_active(tet_id) );
+
+                // Check this won't take us past the max refinement level
+                if (master_element.refinement_level >= MAX_REFINEMENT_LEVEL)
+                {
+                    return false;
+                }
+
+                // If we got here, we didn't detect anything which tells us not
+                // to refine
+                return true;
+            }
+
+            /**
+             * @brief Method which takes a tet id, and deduces the other
+             * parameters needed to perform a 1:2
+             *
+             * @param tet_id The id to refine 1:2
+             */
+            void refine_one_to_two(size_t tet_id)
+            {
+                edge_list_t edge_list = tet_store->generate_edge_keys(tet_id);
+                node_pair_t nodes = find_single_refinement_nodes(edge_list);
```

Comment:
Could use `auto` instead of `node_pair_t`, though some might prefer to see the explicit type.